src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs

```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
                 IgnoreWhitespace = true
             };
 
+            XmlConfigurationElement? root = null;
+
             using (XmlReader reader = decryptor.CreateDecryptingXmlReader(stream, readerSettings))
             {
-                var prefixStack = new Stack<string>();
-
-                SkipUntilRootElement(reader);
-
-                // We process the root element individually since it doesn't contribute to prefix
-                ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                // keep track of the tree we followed to get where we are (breadcrumb style)
+                var currentPath = new Stack<XmlConfigurationElement>();
 
                 XmlNodeType preNodeType = reader.NodeType;
+
                 while (reader.Read())
                 {
                     switch (reader.NodeType)
                     {
                         case XmlNodeType.Element:
-                            prefixStack.Push(reader.LocalName);
-                            ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                            ProcessAttributes(reader, prefixStack, data, AddAttributePair);
-
-                            // If current element is self-closing
-                            if (reader.IsEmptyElement)
                             {
-                                prefixStack.Pop();
+                                var element = new XmlConfigurationElement(reader.LocalName, GetName(reader), GetLineInfo(reader));
+
+                                XmlConfigurationElement parent = currentPath.Any() ? currentPath.Peek() : null;
+
+                                if (parent == null)
+                                {
+                                    root = element;
+                                }
+                                else
+                                {
+                                    if (parent.Children == null)
+                                    {
+                                        parent.Children = new List<XmlConfigurationElement>();
+                                    }
+                                    else
+                                    {
+                                        // check if this element has appeared before, elements are considered siblings if their element names match
+                                        XmlConfigurationElement sibling = parent.Children.FirstOrDefault(e => element.IsSiblingOf(e));
```

Comment:
Let's avoid using LINQ, for perf reasons there is a PR removing LINQ usages here: https://github.com/dotnet/runtime/pull/47873/files#diff-aae949242d873fc3ce6c6e560cc62149d3ecb5f65929936795a812bbc2f950a7R71

Reply:
Count() is LINQ. You want Count, the property.

Reply:
I've added a method XmlConfigurationElement.FindSiblingInChildren that uses a simple for loop.
This also allowed me to make IsSiblingOf a private method.